Cleanup fix

JoshuaRotimi opened the pull request:
This PR fixes the issue #64 with creating multiple instances of the `widget` component across renders
 - Adds a cleanup function to the `uploadWidget` component
 - States a warning in the console when the required env variables are missing to prevent confusion to users on why the Widget might be showing an endless spinner
 - Stores the widget instance in a ref 

These fixes were only applied on the React preset widget example. If all works well, they can be replicated for the Remix and Next.js examples. 

Fixes #64

colbyfayock commented:
hey @JoshuaRotimi this definitely works as expected! i'm going to poke around early next week with this copy and pasted into a project with routing to see how the cleanup / variables are handled to make sure it's handling that use case before merging it in

JoshuaRotimi commented:
Alright sounds great @colbyfayock. After your testing if all works well maybe we can implement the same for the other react centric examples before merging ?

JoshuaRotimi commented:
Alright great! I will make the changes and update the PR

colbyfayock commented:
hey @JoshuaRotimi id be curious if we have this issue with our Next Cloudinary library: https://github.com/colbyfayock/next-cloudinary/blob/main/next-cloudinary/src/components/CldUploadWidget/CldUploadWidget.tsx

we're not destroying the widget instance anywhere, however i dont see a similar issue navigating around page to page like we saw in this example

if you have any interest in playing with that and seeing if there are any issues, it could be another contribution opportunity

there's also a CldVideoPlayer component which I think has a bug open for a similar issue: https://github.com/colbyfayock/next-cloudinary/issues/279#issuecomment-1762134067

no pressure, but thought id share :)

JoshuaRotimi commented:
Yes @colbyfayock , I just tested it out on the Next Cloudinary library and the same issue occurs. Perhaps you can create the issue and assign it to me and I will make the PR to fix it? As for the CldVideoPlayer component bug, I will check it out too.

colbyfayock commented:
assigned you the ticket, thanks Josh! Let me konw when this current PR is ready to review again :)

JoshuaRotimi commented:
Okay great! I will check it out @colbyfayock . Meanwhile This PR is ready to be reviewed again.

colbyfayock commented:
looks like you're trying to commit an `.idea` directory, can you delete that then add `.idea` to the root `.gitignore`? you'll have to do it in separate commits, first deleting, then adding to `.gitignore`

colbyfayock commented on examples/remix-upload-widget-preset/app/components/UploadWidget/index.jsx:
```diff
@@ -39,9 +43,17 @@ function UploadWidget({ children, onUpload }) {
         // "unsigned" uploads which may allow for more usage than intended. Read more
         // about unsigned uploads at: https://cloudinary.com/documentation/upload_images#unsigned_upload
+
+        const cloudName = ENV.CLOUD_NAME;
```
this should be `CLOUDINARY_CLOUD_NAME`

colbyfayock commented:
i'm getting an error in `nextjs-upload-widget-preset` and `nextjs-upload-widget-signed` - did you happen to test these?

<img width="1565" alt="image" src="https://github.com/cloudinary-community/cloudinary-examples/assets/1045274/40ef6a70-6f02-49e3-b86d-4a004b958603">

<img width="1536" alt="image" src="https://github.com/cloudinary-community/cloudinary-examples/assets/1045274/52016b0d-bb0d-4cbc-a1e8-995541c74a33">

JoshuaRotimi commented:
You were very correct! @colbyfayock . Seems like calling the `handleOnLoad` function in the useEffect caused the errors in the images above. I've removed them and also fixed the other issues mentioned. I believe you can test this PR again now. Also makes me wonder why that issue fixed the video rendering on the `CldVideoPlayer` component in the next-cloudinary example. I will take a closer look at that and let you know what I find!

colbyfayock commented:
looks great @JoshuaRotimi thanks for the fixes!

colbyfayock commented:
@all-contributors please add @JoshuaRotimi for code

colbyfayock commented:
@JoshuaRotimi congrats on the merged PR! i can't remember if this is your first hacktoberfest contribution, but if it is your first, since it was merged during Hacktoberfest, you're eligible for some free swag. Please email [hacktoberfest@cloudinary.com](mailto:hacktoberfest@cloudinary.com) with your Github username and a link to this PR where we'll follow up with you there.

https://cloudinary.com/blog/hacktoberfest-celebrate-open-source-sdks

JoshuaRotimi commented:
Thanks @colbyfayock This is my second PR so I've done that already. Always happy to contribute!
